schema/pgsql/upgrades/030.sql

```diff
@@ -0,0 +1,15 @@
+ALTER TABLE rotation ALTER COLUMN mode SET DEFAULT '24-7';
+ALTER TABLE incident ALTER COLUMN severity SET DEFAULT 'ok';
+ALTER TABLE incident_contact ALTER COLUMN role SET DEFAULT 'recipient';
+ALTER TABLE incident_history ALTER COLUMN type SET DEFAULT 'opened';
+
+ALTER TABLE available_channel_type ALTER COLUMN type TYPE varchar(255);
+ALTER TABLE channel ALTER COLUMN type TYPE varchar(255);
+ALTER TABLE contact ADD CHECK (length(username) <= 254);
+ALTER TABLE contact_address ALTER COLUMN type TYPE varchar(255);
+ALTER TABLE object_id_tag ALTER COLUMN tag TYPE varchar(255);
+ALTER TABLE object_extra_tag ALTER COLUMN tag TYPE varchar(255);
+ALTER TABLE browser_session ADD CHECK (length(username) <= 254);
+
+DROP INDEX idx_browser_session_username_agent;
+CREATE INDEX idx_browser_session_username ON browser_session (username);
```

Comment:
I'm afraid browser_session.user_agent is still too large, so I had to shorten the index. @nilmerg @ncosta-ic

Reply:
Well you can't just drop our index on PostgreSQL's side to make it portable for MariaDB. You would need to find a proper compromise.
Now, here are the limits of both database implementations:
- PostgreSQL has technically no limit on its index sizes, however, the standard access method using btree's caps out at a maximum length of 2704 bytes[^1].
- MariaDB has other limitations caused by the underlying storage engine. Let's assume we use InnoDB, as that's the default for MariaDB / MySQL databases. The documentation about indexes states the following[^2]:
> A primary key is **_unique_** and can never be null.
>...
> In [InnoDB](https://mariadb.com/kb/en/innodb/) tables, **_all indexes contain the primary key as a suffix_**. Thus, when using this storage engine, keeping the primary key as small as possible is particularly important. If a primary key does not exist and there are no UNIQUE indexes, InnoDB creates a 6-bytes clustered index which is invisible to the user.

And if we check the information about unique indexes, it states this[^3]:
> ...
> MariaDB starting with [10.5](https://mariadb.com/kb/en/what-is-mariadb-105/)
> Unique, if index type is not specified, **_is normally a BTREE index_** that can also be used by the optimizer to find rows. **_If the key is longer than the max key length for the used storage engine and the storage engine supports long unique index, a HASH key will be created._** This enables MariaDB to enforce uniqueness for any type or number of columns.

Considering the fact that the only reason for a combined index of `username` and `user_agent` would be to find exact matches, a hash index would be more than sufficient. However, that one is apparently not supported with the InnoDB storage engine[^4].
Thus, I would suggest we change the logic again and just store a hash value (Jenkins One-At-A-Time[^5]) with a very small collision rate of the supplied user-agent, like we did in the first iteration of our php daemon code.

I would need the approval of @julianbrost or @nilmerg though.

[^1]: https://dba.stackexchange.com/a/69164
[^2]: https://mariadb.com/kb/en/getting-started-with-indexes/#primary-key
[^3]: https://mariadb.com/kb/en/getting-started-with-indexes/#unique-index
[^4]: https://mariadb.com/kb/en/storage-engine-index-types/#content
[^5]: https://en.wikipedia.org/wiki/Jenkins_hash_function#one_at_a_time , http://burtleburtle.net/bob/hash/doobs.html#statistics

Reply:
<sup>I know, but he can.</sup>

Jokes aside, what if browser_session.user_agent used a single-byte collation (ASCII) and was slightly < 3KB large?

Reply:
Have you considered [using a prefix of the column for the index](https://dev.mysql.com/doc/refman/5.7/en/create-index.html#create-index-column-prefixes)?

In general, how much need is there for such an index? How many sessions may reasonably exist per user? Unless the table is never cleaned up and accumulates rows over the years, iterating over something like 10 sessions of a user shouldn't hurt.

> Considering the fact that the only reason for a combined index of `username` and `user_agent` would be to find exact matches, a hash index would be more than sufficient.

What will this index be used for in the end? Like I'd have expected this to be used for some "my sessions" overview where you'd do a `select ... where username = '...' order by user_agent`. Shouldn't the primary way of selecting a specific session be the session ID anyways?

Reply:
> Have you considered [using a prefix of the column for the index](https://dev.mysql.com/doc/refman/5.7/en/create-index.html#create-index-column-prefixes)?

Good idea! Especially this would enable un-limiting the UA length.

Our current 4k is nice, but there is no limit actually: https://stackoverflow.com/a/77072435

Header limits:

* Apache: 8k https://lists.apache.org/thread/vq8z74zl0tsb887n9t7y8f3fo7kp3klx
* nginx 8k: https://nginx.org/en/docs/http/ngx_http_core_module.html#:~:text=A%20request%20header%20field%20cannot,is%20equal%20to%208K%20bytes.

This means I could send you a 7.9k UA.

> In general, how much need is there for such an index? How many sessions may reasonably exist per user? Unless the table is never cleaned up and accumulates rows over the years, iterating over something like 10 sessions of a user shouldn't hurt.

Exactly! Even if MySQL is not smart enough to use the 1-column index when you filter/sort by 2 columns, one can filter/sort by 1 column and do the rest in userspace.

> > Considering the fact that the only reason for a combined index of `username` and `user_agent` would be to find exact matches, a hash index would be more than sufficient.

Apropos hashing, Eric would just store the indexed UA as hash...

> What will this index be used for in the end? Like I'd have expected this to be used for some "my sessions" overview where you'd do a `select ... where username = '...' order by user_agent`. Shouldn't the primary way of selecting a specific session be the session ID anyways?

... but I'd (at least additionally) include the unlimited(!) UA for overview purposes.

Reply:
> In general, how much need is there for such an index? How many sessions may reasonably exist per user? Unless the table is never cleaned up and accumulates rows over the years, iterating over something like 10 sessions of a user shouldn't hurt.
> 
> What will this index be used for in the end? Like I'd have expected this to be used for some "my sessions" overview where you'd do a `select ... where username = '...' order by user_agent`. Shouldn't the primary way of selecting a specific session be the session ID anyways?

Well, to be fair, the only reason that we included the UA to begin with, is to be able to differentiate between multiple valid user sessions. User `A` could be logged in on multiple devices at the same time and should thus receive the desktop notifications on all of them.

Now one might think that the unique PHP session identifiers would be sufficient for this, but there's one problem that we can't really solve in a neat way: We don't know when PHP cleans up a session, as this is dependent on the PHP configuration and on how frequent requests towards the web server are being made. And, one could just close the tab or browser instead of using the logout function, which completely bypasses Icinga's session handling (that we are in control of).

Because of this, we included a cleanup job into our PHP daemon, which cleans up sessions in the database that have been authenticated for more than 24 hours.
This, however, means, that PHP could technically assign the same identifier to two different sessions within a timespan of 24 hours.

We trust PHP's session uniqueness in behalf of its implementations and clean up a (potentionally) existing session in our table, if there's a new, successful login that uses the same identifier. The reason being, that this could only ever happen, if we would still have a zombie session saved to our database, that PHP already cleaned up.

To keep our zombie session count low, the PHP daemon checks for [a potential zombie session on a successful login](https://github.com/Icinga/icinga-notifications-web/blob/ba627b479f02b838fa83151cb8b787dd5af36f5b/library/Notifications/ProvidedHook/SessionStorage.php#L71-L86). This check provides the user-agent, since we want to preserve other valid sessions of that user that are tied to other devices. If we did not include the user-agent, this check would clean up all other sessions of that user, even the ones which are still in use.

I hope that gives you a better idea about the reasoning behind this index and why it's needed. I am still sure, that a simple hash over the complete user-agent would be more than sufficient for this specific demand, but @Al2Klimov has a valid point. For the time being, the desktop notifications feature is the only part of `icinga-notifications-web` that makes use of it.
As for the future, we do plan to merge the PHP daemon with additional logic to the Icinga2 (Web) core. This could then create additional demands, which would want to differentiate between different aspects of a connection (like the browser, operating system, you name it...). If we keep this in mind, there is a reason to store the full user-agent into the database. However, it would not need to be indexed.

Reply:
> If it does, it would probably be a security issue in PHP. That would likely imply that you can guess session IDs because they are not random enough.

You're right. I glanced over the current PHP implementation and it seems like they properly solved it for newer systems and installations:
[session.c](https://github.com/php/php-src/blob/9d3907fd8564b47e8d7e9ebf68161ac12ceb0455/ext/session/session.c#L341-L346) > [csprng.c](https://github.com/php/php-src/blob/9d3907fd8564b47e8d7e9ebf68161ac12ceb0455/ext/random/csprng.c#L69)
For systems without proper csprng support or on a PHP version under 7.1[^1], they still fallback to the old sha1 method:
[php_random.h](https://github.com/php/php-src/blob/9d3907fd8564b47e8d7e9ebf68161ac12ceb0455/ext/random/php_random.h#L45-L47) > [random.c](https://github.com/php/php-src/blob/9d3907fd8564b47e8d7e9ebf68161ac12ceb0455/ext/random/random.c#L685-L687) > [random.c](https://github.com/php/php-src/blob/9d3907fd8564b47e8d7e9ebf68161ac12ceb0455/ext/random/random.c#L613-L630)

> That looks like something where you try to clean up some sessions even before their expiry after 24 hours, i.e. other sessions of the same user that share the same user agent string. I feel like I've said this before, but I wouldn't make the assumption that the user agent string uniquely identifies a user's device. I currently have three devices in this room that will report the very same Chrom(e|ium) on Linux user agent.
> 
> To be honest, that sounds like an optimization that should not be necessary and might even cause issues in some scenarios. If a user signs in five times a day on the same device, would it cause any trouble if those entries in the table would live for a day?

No, not really. I was just trying to follow the session security guidelines by PHP[^2], which state:
> Developers should keep track of all active sessions for every user.
> ...
> Various ways to implement this exist. One possible implementation is setting up a database that keeps track of the required data and stores any relevant information. Since session data is GCed, developers must take care of the GCed data to maintain the active session database consistency. 
Meaning if we don't clean them up (when they have been garbage collected by PHP itself), a session id could still be used to connect to the PHP daemon. This could leak check data to an unwanted user.

This lead me to the idea of manually cleaning up (confirmed) zombie sessions.

> I mean if you think it's necessary, you can have it. To me, solutions that don't require external workarounds for the database sound preferable. And truncating the column for the index sounds like a fair compromise where you still get the help of the index as when the first N characters of the user agent differ (which they typically should if N is chosen sufficiently large so that it extends beyond the useless `Mozilla/5.0` prefix that all browsers send).

I'm with you on that one. Albeit, I don't really think that calculating a hash with the native PHP functions would be considered a workaround, as the only thing it's supposed to do, is, to give information about the equality of two user agents. And yes, just reading a user-agent obviously won't be a 100% solution to detect different devices. But I really cannot condone *implementing fingerprinting* - even for such a feature. That's a no, no...

[^1]: https://www.php.net/manual/en/features.session.security.management.php#features.session.security.management.non-adaptive-session comment about `session_create_id()` and its note.
[^2]: https://www.php.net/manual/en/features.session.security.management.php#features.session.security.management.active-sessions